src/java.base/share/classes/jdk/internal/jimage/ImageReader.java

```diff
@@ -381,18 +500,29 @@ Node findResourceNode(String moduleName, String resourcePath) {
          *
          * <p>This method is expected to be called frequently for resources
          * which do not exist in the given module (e.g. as part of classpath
-         * search). As such, it skips checking the nodes cache and only checks
-         * for an entry in the jimage file, as this is faster if the resource
-         * is not present. This also means it doesn't need synchronization.
+         * search). As such, it skips checking the nodes cache if possible, and
+         * only checks for an entry in the jimage file, as this is faster if the
+         * resource is not present. This also means it doesn't need
+         * synchronization most of the time.
          */
         boolean containsResource(String moduleName, String resourcePath) {
             if (moduleName.indexOf('/') >= 0) {
                 throw new IllegalArgumentException("invalid module name: " + moduleName);
             }
-            // If the given module name is 'modules', then 'isResource()'
-            // returns false to prevent false positives.
+            // In preview mode, preview-only resources are eagerly added to the
+            // cache, so we must check that first.
+            if (previewMode) {
+                String nodeName = MODULES_PREFIX + "/" + moduleName + "/" + resourcePath;
+                // Synchronize as tightly as possible to reduce locking contention.
+                synchronized (this) {
+                    Node node = nodes.get(nodeName);
+                    if (node != null) {
+                        return node.isResource();
+                    }
+                }
+            }
             ImageLocation loc = findLocation(moduleName, resourcePath);
-            return loc != null && isResource(loc);
+            return loc != null && loc.getType() == RESOURCE;
```

Comment:
Would there be any benefit to caching the resource here, since it has been found and will likely be opened again.